connectivity/connectivity.go

```diff
@@ -61,3 +60,32 @@ const (
 	// Shutdown indicates the ClientConn has started shutting down.
 	Shutdown
 )
+
+// ServingMode indicates the current mode of operation of the server.
+//
+// Only xDS enabled gRPC servers currently report their serving mode.
+type ServingMode int
```

Comment:
LGTM.

---

This actually promotes an xDS only concept to a gRPC concept.

The concern is that in the future we may change the meanings of these terms in a non-xDS gRPC server. But with the 3 values we have now, that seems unlikely.
- non-xDS server doesn't have `starting` with our current API. If our API is to take a port instead of a listener, the state after `NewServer`, before the `net.Listen()` would be `ServingModeStarting`
- serving is serving
- `NotServing` is after gracefulstop, before stop?

Reply:
> * non-xDS server doesn't have `starting` with our current API. If our API is to take a port instead of a listener, the state after `NewServer`, before the `net.Listen()` would be `ServingModeStarting`

Even with our current API taking a `net.Listener`, we could have a `starting` state after `NewServer` and before we call `Accept()`. It is very fleeting though.

> * `NotServing` is after gracefulstop, before stop?

Sounds about right.

Reply:
> Sounds about right.

What happens at "stop", then?  It no longer has any serving mode?  I'd think the `NotServing` state continues after "stop", too.

Reply:
We have a starting. So should we also add a `stopped` (definitely not stopping).
- NotServing means no ___new___ streams?
- Stopped means no streams?

Or we can add them when needed.

Reply:
> > Sounds about right.
> 
> What happens at "stop", then? It no longer has any serving mode? I'd think the `NotServing` state continues after "stop", too.

The only difference I wanted to highlight was that in `NotServing` mode, we accept and close the connection without performing any I/O (this is a limitation of our API, since we accept a `net.Listener`). Once stop is complete, we don't even accept the connection.

Reply:
> NotServing means no new streams?

Or "no new _connections_" is what I thought.  Existing connections may allow new streams.

Reply:
"no new streams" and "no new connections" are different.
GracefulStop also sends GOAWAYs, so it won't accept new streams.

`completely-NotServing` vs `almost-NotServing`? :see_no_evil: